src/main/java/com/thealgorithms/strings/AhoCorasick.java

```diff
@@ -0,0 +1,228 @@
+
+                    // Check if the current node has a child for the current character
+                    if (curr.getChild().containsKey(c)) {
+                        curr = curr.getChild().get(c); // Update the current node to the child node
+                    } else {
+                        // If no child node exists, create a new one and add it to the current node's children
+                        Node nn = new Node();
+                        curr.getChild().put(c, nn);
+                        curr = nn; // Update the current node to the new child node
+                    }
+                }
+                curr.setPatternInd(i); // Store the index of the pattern in the current leaf node
+            }
+
+            // Build suffix and output links after constructing the trie
+            root.setSuffixLink(root); // Initialize the suffix link of the root to itself
+            Queue<Node> q = new LinkedList<>(); // Initialize a queue for BFS traversal
+
+            // Initialize suffix links for child nodes of the root
+            for (char rc : root.getChild().keySet()) {
+                Node childNode = root.getChild().get(rc);
+                q.add(childNode); // Add child node to the queue
+                childNode.setSuffixLink(root); // Set suffix link to the root
+            }
+            while (!q.isEmpty()) {
+                Node currentState = q.poll(); // Get the current node for processing
+
+                // Iterate through child nodes of the current node
+                for (char cc : currentState.getChild().keySet()) {
+                    Node currentChild = currentState.getChild().get(cc); // Get the child node
+                    Node parentSuffix = currentState.getSuffixLink(); // Get the parent's suffix link
+
+                    // Calculate the suffix link for the child based on the parent's suffix link
+                    while (!parentSuffix.getChild().containsKey(cc) && parentSuffix != root) {
+                        parentSuffix = parentSuffix.getSuffixLink();
+                    }
+
+                    // Set the calculated suffix link or default to root
+                    if (parentSuffix.getChild().containsKey(cc)) {
+                        currentChild.setSuffixLink(parentSuffix.getChild().get(cc));
+                    } else {
+                        currentChild.setSuffixLink(root);
+                    }
+
+                    q.add(currentChild); // Add the child node to the queue for further processing
+                }
+
+                // Establish output links for nodes to efficiently identify patterns within patterns
+                if (currentState.getSuffixLink().getPatternInd() >= 0) {
+                    currentState.setOutputLink(currentState.getSuffixLink());
+                } else {
+                    currentState.setOutputLink(currentState.getSuffixLink().getOutputLink());
+                }
+            }
+        }

```

Comment:
This should be cut into smaller pieces.

Reply:
> * one method `searchIn` taking the `text`.
>   The constructor would do the same as `buildTrie` and `buildSuffixAndOutputLinks`. The `searchIn` would return the data, which is currently stored in `res`.

Ah yes, It might be my misunderstanding because when I read the above I thought you need only one function in  `Trie class` that is `searchIn` and  let `constructor` do the work of the `buildTrie` and `buildSuffixAndOutputLinks` and merged them into constructor.

I will revert them back, if there's still any issue please let me know.

Reply:
Constructors can also call other methods.